### PR TITLE
Modeling doc reorganization

### PR DESCRIPTION
**Pull request by karllark**
First part to reorganized, update, and enhance the modeling documentation.  This PR is to do the reorganization - mainly moving around the existing content so that it is easier to find information and support the planned expansion of the documentation.  The plan is for other PRs to update/enhance the modeling docs.

A rendered version of the docs can be found by clicking on the 'giles' link in the testing summary below.

Partially addresses #9069.

### CODE REVIEW

File: CHANGES.rst [hunk 1]
```diff
@@ -59,6 +59,8 @@ astropy.modeling
   See modeling documentation for details.
   `<https://docs.astropy.org/en/v4.0/modeling/changes_for_4.html>`_ . [#8769]
 
+- Significant reorganization of the documentation.
```
Comment (nden) [inline CHANGES.rst]: You need to add the PR number at the end of the change entry, like [#9078]

Comment (karllark) [reply CHANGES.rst]: Forgot that.  Will do so.

**Comment by karllark** [comment]
Ready for review.

**Comment by bsipocz** [comment]
@karllark - the rendered docs is also available by the "giles" bot for each commit, no need to build it elsewhere, look for the link in the status box at the bottom of the PR.

**Comment by karllark** [comment]
> @karllark - the rendered docs is also available by the "giles" bot for each commit, no need to build it elsewhere, look for the link in the status box at the bottom of the PR.

Very cool.  I did not see this.  So much easier and better than my cached version.  Thanks for pointing this out!

**Comment by hcferguson** [comment]
@karllark 

In general I think the reorganization looks good. A few minor suggestions.
For the first example under getting-started -- I think it would be good to include the code for making the plot, since beginners like to cut & paste and see what is in the docs. I don't know if it's possible to just include a link for the source code (e.g. add a src link in addition to the png, svg, pdf link) rather than rendering it in the document, but in this case it's just a few lines of code to make the plot so it doesn't clutter it too much.

Breaking the document up in to sections on separate pages (rather than the same page with anchors) has its advantages and disadvantages. I think it's okay, but I have a couple worries: (1) will this actually reduce discoverability via google search? When it's all one page, that page becomes popular and gets a higher ranking. When it's split up, the effect is diluted. But I don't really know enough about this for this to be an informed worry. 

Having broken the document up this far, it seems pretty ugly to leave the Reference/API section on the landing page. That's the least beginner-friendly part of the documents. So I'd suggest putting that off in a separate page as well. 

The Model Set discussion would benefit from some high-level motivation (what is the killer use case that drove this)? 

In the same section, it's a bit jarring that the Further Examples that come right after the discussion of model separability aren't illustrating model separability. Seems like these examples would be better up in the Model Set section.

File: docs/modeling/example-fitting-uncertainties.rst [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+                     + c*np.square(np.full(x.shape,5.0)))
+    g_init = models.Gaussian1D(amplitude=1., mean=0, stddev=1.)
+
+    # initialize fitters
+    fit = fitting.LevMarLSQFitter()
+
+    # fit the data w/o weights
+    fitted_model = fit(g_init, x, y)
+
+    # fit the data using the uncertainties as weights
```
Comment (hcferguson) [inline docs/modeling/example-fitting-uncertainties.rst]: Wouldn't one usually use the inverse variance as a weight -- i.e. 1/y_unc**2?

File: docs/modeling/model-sets.rst [hunk 3]
```diff
@@ -0,0 +1,61 @@
+.. include:: links.inc
+
+Model sets
+==========
+
+In some cases it is necessary to describe many models of the same type but with
+different sets of parameter values.  This could be done simply by instantiating
+as many instances of a `~astropy.modeling.Model` as are needed.  But that can
+be inefficient for a large number of models.  To that end, all model classes in
```
Comment (hcferguson) [inline docs/modeling/model-sets.rst]: This section might benefit from a more specific example of an application that would benefit significantly from the Model Set machinery.

**Comment by hcferguson** [comment]
The section on "Defining a Plugin Fitter" is repeated between the Advanced Fitting and the Defining New Fitter Class sections. Not sure if that's intentional.

**Comment by hcferguson** [comment]
There is a Model Sets discussion under Advanced Concepts. But there is also a separate (shorter & less detailed) section called Model Sets.

It seems cleaner to have this in its own section, so that Advanced Concepts could just point to that section. But the current version under Advanced Concepts has some important discussion differentiating model sets from parameters that are arrays that isn't currently in the separate section.

**Comment by perrygreenfield** [comment]
On the basics page it may be worth clarifying in the next to last sentence that input arrays are not broadcast with each other (an exception is for scalar inputs that are broadcast to input arrays). Input arrays are broadcast with parameters though exactly how probably belongs in more advanced topics. Maybe we should just say that models generally take arrays as inputs.

**Comment by perrygreenfield** [comment]
I would tend to split the topics in advanced concepts into separate pages and being shown as list under Advanced Concepts on the landing page. The first section could be relabeled Using arrays for parameters and inputs. 

As for the canonical case for model sets, it really is when many parallel linear fits must be done over many data series that use the same model; a classic example is fitting response for each pixel using a polynomial. It is much more efficient to bundle them together than to loop over all pixels. Seems to me that this makes a good motivating example (It's why we did it!)

**Comment by perrygreenfield** [comment]
In short, perhaps this organization of topics at the top level:

Models:
- Basic Concepts
- Using Arrays as parameters and model variables
- Inverses
- Model Sets
- Input/Output dependencies
Combining Models
- Basics
- Advanced usage
Fitting:
- Basic Fitting
- Advanced Fitting
- Obtaining Fitting Errors (TBD)
- Using a Custom Statistic Function
- Using models with Bayesian tools (TBD)
Using Units with Models and Fitting
Extending Models and Fitting:
- Making a new model from an existing function
- Making a new model from scratch
- Adding or modifying Fitters
Performance tips and general advice
Modeling changes in astropy v4.0

**Comment by jdavies-st** [comment]
I echo Harry's points below:

> Breaking the document up in to sections on separate pages (rather than the same page with anchors) has its advantages and disadvantages.
> 
> Having broken the document up this far, it seems pretty ugly to leave the Reference/API section on the landing page.

The automod_api section is the least useful for users, so having it on the landing page is not great.  Agree with Harry that a single page with anchors would be easier to navigate and would aid discoverability of topics.

Punt the API stuff into it's own separate page if possible, or just leave it at the end as it is now, if you decide to make it all one page.

**Comment by karllark** [comment]
Thanks for all the great comments.  I will work as many as I can into this PR in the very near future.

File: docs/modeling/example-fitting-uncertainties.rst (hunk 2 above)
Comment (karllark) [reply docs/modeling/example-fitting-uncertainties.rst]: Actually, this is the correct way to pass the weights.  According to the documentation for the `LevMarLSQFitter` fitter, 1/unc is what should be passed.

```
        weights : array (optional)
            Weights for fitting.
            For data with Gaussian uncertainties, the weights should be
            1/sigma.
```

File: docs/modeling/model-sets.rst (hunk 3 above)
Comment (karllark) [reply docs/modeling/model-sets.rst]: I agree.  And it should be in a separate PR that focuses on model sets.

**Comment by karllark** [comment]
@hcferguson :  Moved the API to a separate page - good point.  As for the google search optimization, that is something I don't know much about.  But I do feel that all the modeling information one page is just too much material.  Separate pages are needed.  Obviously if we can get the "good" information (according to google) on the main page that is a great thing.  I'm open to suggestions.

As for showing the source code for the example on the index page.  Giving a link to the plot source code is disabled in the astropy conf.py file with the line `plot_html_show_source_link = False` and I cannot find a way to override this.  Instead, I just show the full code for the plot so that everything is there.  A good solution I think.

**Comment by karllark** [comment]
@perrygreenfield : I like the direction that you are going with the reorg of the index page, but I'm not clear on what content we currently have can go in each listing.  I feel like we are missing some of the material you mention (in addition to the TBD sections).  A face-to-face discussion may make it easier to understand things.  In addition, I feel that putting together the better landing page listing (index.rst) might be better done incrementally as the material in each section is revised.

So, I am recommending we merge this PR and then address your nice reorg listing in separate PRs.

**Comment by perrygreenfield** [comment]
That would be fine with me (merging and deferring further partitioning)

**Comment by pllim** [comment]
For future reference, please keep the existing Sphinx linking tags if possible. This broke my doc build downstream that uses Intersphinx to those tags. I'll open a follow-up PR.

**Comment by karllark** [comment]
Ok.  Please tag me in the PR so I can learn exactly what you mean.  I thought I'd kept all those tags, but clearly I missed at least one.

**Pull request merged**